tensorflow/core/kernels/fft_ops.cc

```diff
@@ -97,9 +84,89 @@ class FFTGPUBase : public OpKernel {
   virtual bool IsForward() const = 0;
   virtual bool IsReal() const = 0;
 
- private:
+  // The function that actually computes the FFT
+  virtual void DoFFT(OpKernelContext* ctx, const Tensor& in, uint64* fft_shape,
+                     Tensor* out) = 0;
+};
+
+typedef Eigen::ThreadPoolDevice CPUDevice;
+
+template <typename Device, typename TInput, typename TOutput,
+          int FFTResultType, int FFTDir, int FFTRank>
+struct FFTFunctor {
+  void operator()(const Device& d,
+                  typename TTypes<TOutput, FFTRank + 1>::Tensor output,
+                  typename TTypes<TInput, FFTRank + 1>::ConstTensor input) {
+      // Cast to non-const data type to ensure we can call the template fft
+      // function
+      // TODO(tillahoffmann): Remove reinterpret_cast once Eigen supports const
+      // arguments for tensor fft
+      auto casted_input =
+       *reinterpret_cast<typename TTypes<TInput, FFTRank + 1>::Tensor*>(&input);
+      // Create the axes (which are always trailing)
+      auto axes = Eigen::ArrayXi::LinSpaced(FFTRank, 1, FFTRank);
+      // Evaluate the fft on the specified device
+      output.device(d) = casted_input.template fft<FFTResultType, FFTDir>(axes);
+  }
+};
+
+template <bool Forward, bool _Real, int FFTRank>
+class FFTCPU : public FFTBase {
+ public:
+  using FFTBase::FFTBase;
+ protected:
+  int Rank() const override { return FFTRank; }
+  bool IsForward() const override { return Forward; }
+  bool IsReal() const override { return _Real; }
+
   void DoFFT(OpKernelContext* ctx, const Tensor& in, uint64* fft_shape,
-             Tensor* out) {
+             Tensor* out) override {
+    // TODO(tillahoffmann): Support RFFTs by slicing away the negative
+    // frequency components
+    OP_REQUIRES(ctx, !IsReal(), errors::Internal("Real FFT not supported."));
+    auto input = in.flat_inner_dims<complex64, FFTRank + 1>();
```

Comment:
I'd slightly prefer to avoid the above reinterpret_cast since I'm not sure converting a `TensorMap<const T, ...>*` to a `TensorMap<T, ...>*` will always be undefined-behavior free. Maybe someone more expert than me at interpreting the standard or Eigen can chime in :).

Since Tensor types are lightweight wrappers around reference-counted buffers, you can [copy](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/tensor.h#L161
) away the const-ness cheaply.
```c++
Tensor nonconst_in = in;
auto input = nonconst_in.flat_inner_dims<complex64, FFTRank + 1>();
```